### ISSUE TITLE
Ideas from Minigo

### ISSUE THREAD

**Issue by gcp** [leela-zero/leela-zero#785]
https://github.com/tensorflow/minigo

The TensorFlow team did their own re-implementation of AlphaGo Zero, apparently meant to demonstrate TensorFlow and how to run machine learning apps on the Google Cloud. I had a peek at how they resolved some things which were considered ambiguous in the AlphaGo Zero paper:

They use FPU=parent, not FPU=0 (in -1...1 range):
https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/mcts.py#L204

They added specific knowledge about passes to avoid the early pass wars (boo, cheating!): 
https://github.com/tensorflow/minigo/commit/a8af0d8280b4b80a53ed92e59428d1990cbb3a46

They store the probabilities that were used to pick the move (not the original search probabilities)
https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/strategies.py#L135
https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/strategies.py#L233

but it's not quite T=0 vs T=1:
https://github.com/tensorflow/minigo/blob/a3e98185c192636389aff113e674a2cf8e0d9488/mcts.py#L223

**Comment by gcp** [comment]
I think this is a great idea, I'd pick C++ over Go though.

https://github.com/tensorflow/minigo/issues/14

**Comment by gcp** [comment]
Scale/Center layers used for the batchnorm during residual:
https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/dual_net.py#L213

But not for the head:
https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/dual_net.py#L237

(Scale is entirely redundant, but Center may not be - it's one of the things I'd fix in a v2 weights format)

I think it also has the problem that the convolution bias weights are entirely redundant.

**Comment by gcp** [comment]
They also batch the MCTS search by 8:
https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/strategies.py#L177

I'm guessing the loss in search efficiency is worth it compared to any speedups they get due to the backend being TF+cuDNN.

**Comment by Ttl** [comment]
One small difference I noticed that in minigo L2 regularization is not applied to the bias terms. https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/dual_net.py#L268

Reading around it seems that generally it's recommended not to apply regularization to biases since large biases don't cause overfitting in the same way as having many non-zero weights. No idea how significant it is though.

**Comment by gcp** [comment]
Most of our bias weights are also excluded because they're marked as not trainable, and non-trainable variables don't get regularization applied.

It would affect the heads, though. Easy to change that one.

**Comment by justinwong098** [comment]
It's interesting that even though the TensorFlow team is also owned by Google, that they don't necessarily have insider information into how AlphaGo was built, outside of what was described in the paper (although I guess it makes sense, since they are separate arms under Alphabet).

They talk about some of their mistakes/tweaks/adjustments [here](https://github.com/tensorflow/minigo/blob/master/RESULTS.md).

**Comment by gcp** [comment]
>They talk about some of their mistakes/tweaks/adjustments here.

That's very interesting reading. The parallels to the experiences we had (and the same doubts about some of the ambiguities!) are very clear.

"Around generation 95, we realized the initial Q-value led to weird ping-ponging behavior inside search, and changed to init to the average of the parent."

"we started experimenting with tree reuse to gather data about its effect on the Dirchlet noise"

"here's still ambiguity about whether or not a onehot training target for the policy net is good or bad"..."Ambiguity around the transformation of the MCTS output, pi."

"This suggested our new dataset code was not shuffling adequately, leading to value overfitting"

"Figure out an equivalent learning rate decay schedule to the one described in the paper."

**Comment by justinwong098** [comment]
[Here](https://www.reddit.com/r/cbaduk/comments/7tx4qm/an_opensource_implementation_of_the_alphagozero/dth6p2w/) is some more information about this project's background.

>we're not affiliated with the deepmind team at all, correct! We're also an enthusiast project -- that 20% time you hear about -- and we're not actually part of tensorflow. They have been very generous in answering questions and offering to host our code on their repo, which greatly sped up our ability to get this public!
> The project is also a great demonstration of the latest TF apis -- MuGo originally used the earlier TF api, and we recently updated it to use Datasets, Layers, all the new fun stuff.

**Comment by amj** [comment]
Hi!  We're not actually part of the tensorflow team -- just a group of Go players working on 20% time and persuading people it was worth doing as OSS.

We're pretty close to feature parity with the AGZ paper and we're hoping to make another run at the summit, wiping and starting over.  We'd love to have some extra eyes on the code and in particular, i'm hoping that putting our heads together on some of these ambiguities can be fruitful for both of us.

Let me address a few of these things:
 - Batch by 8.  Earlier games used root parallelization [described here](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.159.4373&rep=rep1&type=pdf), i.e., we played 8 games in parallel, 8x slower.  There were no 'virtual losses' enabling us to select different nodes from the same tree repeatedly.  However, it might be possible that the vlosses essentially 'broaden' the search in a meaningful way.  So we've got it in now.
- Temperature.  We used 1/8th there for our first run, but we've recently reverted that to unity temperature always. We tried 'squashing' the probabilities with t>1 on the latest in our attempt to get it to stop playing the 3-3 point....hasn't worked yet ;)
- We recently rewrote the whole dataset code to use Tf.Dataset and not the MuGo Dataset.  Prior to that, we had a very arduous, very slow, very crappy process to select uniformly from the most recent 125M positions (not 500k games).  the new re-write hasn't quite had the effect we were hoping on.
- Our early passes were actually even worse -- we'd turned off pass for the first ~80ish moves.  (it was some arithmetic function that would work for boardsize 9 as well i forget what).
- Batchnorm: i still have doubts that i have this correct :(

So before we take our next run at the summit we'd like to do these:
 - Add rotation in training
 - Upgrade to py3.6, TF 1.5, cuda 9
 - figure out the equivalent decay rate -- our intuition was that smooth decay would be strictly better than stepwise, but, maybe not?
 - possibly also revisit how we balance training & selfplay.  We aren't doing any evaluation, AZ style (this was originally driven by being even more resource constrained, then the AZ paper hit arxiv and i felt clever ;)  But it's possible the evaluation of new networks was the way they figured out how to keep selfplay & training in balance, and once they figured that out they dropped it in AZ.  Either way, our pipeline is real dumb about this: if the training job dies or slows down, we get more games in the current model, if the cluster has a lot of pre-emptions, a generation will have fewer games.

Really happy to finally be able to chat about these things with everyone :)  It's been hard watching these really great conversations and not being able to participate!  :)

**Comment by Ttl** [comment]
> (Scale is entirely redundant, but Center may not be - it's one of the things I'd fix in a v2 weights format)

If you really want it's possible to add beta in training and use the same weight format by modifying mean and variance terms when writing the weight file.

With substitution `s = 1/sqrt(var + ε)` batch norm update with beta is: `y = s * (x - µ) + beta`. This can be rewritten to include beta in µ as: `y = s * (x - (µ - beta / s)) = s * (x - µ')`, where `µ' = µ - beta * sqrt(var + ε)`.

**Comment by gcp** [comment]
>If you really want it's possible to add beta in training and use the same weight format by modifying mean and variance terms when writing the weight file.

Very clever (as usual :P).

This might be worth testing with a supervised run over existing data, i.e. whether adding the Center layer gives a lower eventual loss.

**Comment by gcp** [comment]
>We tried 'squashing' the probabilities with t>1 on the latest in our attempt to get it to stop playing the 3-3 point....hasn't worked yet ;)

FWIW LZ is currently pretty locked to 4-4, but:

```
Leela: heatmap 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   2  38   0   0   0   0   0   0   0   0   0   0   0  31   2   0   0 
  0   0  35 135   1   0   0   0   0   0   0   0   0   0   1 130  29   0   0 
  0   0   0   2   0   0   0   0   0   0   0   0   0   0   0   2   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0 
  0   0  32 149   1   0   0   0   0   0   0   0   0   0   2 145  29   0   0 
  0   0   2  35   0   0   0   0   0   0   0   0   0   0   1  41   2   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 
pass: 0
```

I believe that in long searches it does sometimes ends up playing 3-4. So maybe it's not needed to force the issue.

**Comment by gcp** [comment]
It would be really valuable if network weights files were exchangeable between the two projects, right?

**Comment by isty2e** [comment]
@gcp But aren't the numbers of input features different?

**Comment by gcp** [comment]
Hmm true. I wonder how @amj feels about the issue with "only black knows about the board edge" and if they want to make it compatible.

With black to move it may be possible to make the networks equivalent, but I don't see how to do it with white.

**Comment by gcp** [comment]
In any case it would be quite to adapt LZ to take a miniGo format network, just detect a different network version and change the input features.

**Comment by jkiliani** [comment]
Could training data be interchanged? It wouldn't be strict self-play of course, but maybe it could actually be useful for bootstrapping?

**Comment by gcp** [comment]
>Could training data be interchanged?

Don't see why not.

**Comment by amj** [comment]
It's certainly interesting!  i think we would certainly be able to see if there are differences in the harness code if we could pass weights back and forth.  It'd probably make testing easier and establish some good baselines.  I'm open to it assuming we can figure it out :)

Re: the 'board edge' change -- i honestly don't know enough about it to have an opinion.  It seems like the inputs could be translated easily enough?  It'd be a little different than 'hold model steady, vary the harness', but it would still be pretty interesting.

Re: training data, yeah totally, with the caveat that a lot of our data is crap ;)  I'm still working on how to host/serve the data and find it a home -- i definitely want it all to be out there and open! -- so we'll figure out a way to get that up soon.

**Comment by gcp** [comment]
>I'm still working on how to host/serve the data and find it a home -- i definitely want it all to be out there and open! -- so we'll figure out a way to get that up soon.

FWIW our storage was sponsored by online-go.com (see https://github.com/gcp/leela-zero/issues/167) and the backing provider is...drumrolll...Google Cloud.

**Comment by Ttl** [comment]
LZ can use minigo weights if the weights of the last missing plane in input convolution are set to zero. However using LZ weights in minigo doesn't seem possible without code changes or training. Although minigo can add the white to move plane any time and use the same weights by setting the weights of the extra plane to zero, so it isn't that big modification.

**Comment by marcocalignano** [comment]
@amj  Hoe are you doing the self playing, distributed like us or you have a dedicated hardware?
I guess it is also possible to generate new training  playing Leelaz against miniGo.
I read that maybe you want to do the self playing in another language. If you choose to do it in C++ I would like to help. We could start little using PyQt to use the python engine in the beginning and go faster from there.

**Comment by gcp** [comment]
@marcocalignano They are using (sponsored - it's their employer) Google Cloud instances with about 1000 nodes per run.

**Comment by brilee** [comment]
To add to the "only white knows about edge of the board", I think one of the ways in which this may matter less is that if one of the filters in the first layer manages to train its output to be all ones, then this would allow subsequent layers to treat the first layer filter as a ones plane. Since it's possible in theory, we wave our hands, say "machine learning" and then it just works.

**Comment by roy7** [comment]
> I believe that in long searches it does sometimes ends up playing 3-4. So maybe it's not needed to force the issue.

My 1600 playout LZ bot always goes 4-4, but my slower 50000 playout LZ bot often goes with 3-4. Some of the first 6x128 bootstrap networks used odd openings but that seems behind us now.

**Comment by wctgit** [comment]
>> Could training data be interchanged?
>
> Don't see why not.

Well, if you want to stick to the root principle of AGZ, then borrowing training data from Minigo would be importing some human prior knowledge, specifically of the passing tweak.

**Comment by odeint** [comment]
I guess it would be even cleaner to train yet another AI to generate the network architecture, hyperparameters and training schedule. It would remove the awful human influence one layer further - maybe play will become stronger, still!

**Comment by alreadydone** [comment]
[People who thought somebot was Leela Zero](https://tieba.baidu.com/p/5500458833) were not too far from truth :)

**Comment by l1t1** [comment]
lz is stronger than somebot 199b at http://www.yss-aya.com/cgos/19x19/bayes.html now 

469 | somebot-199b | 2606 | 62 | 62 | 237 | 50% | 2690 | 0%
-- | -- | -- | -- | -- | -- | -- | -- | --


448 | LZ-879ad1-t1-p1600 | 2668 | 94 | 94 | 113 | 51% | 2713 | 0%
-- | -- | -- | -- | -- | -- | -- | -- | --

**Comment by Ttl** [comment]
I ran a training with and without center option in all batch_norms. I used net_to_model with 6x128 supervised trained weights for initialization. Batch size was 128, learning rate was at first 0.0005, but since loss didn't seem to improve after 10k steps I dropped it to one tenth. Training data was all human games from https://github.com/yenw/computer-go-dataset plus 7d+ games from https://u-go.net/gamerecords/. Test set was one chunk of 4d games from https://u-go.net/gamerecords-4d/.

I modified tfprocess to print training results every 100 steps and test results with 100 batches every 200 steps. Below are the training loss graphs:

![center_train](https://user-images.githubusercontent.com/544240/35681742-03d784f2-0767-11e8-9f65-9d9574f8ea32.png)
![center_test](https://user-images.githubusercontent.com/544240/35681744-05739044-0767-11e8-9fdd-c18d56f60896.png)

Network with center=True has slightly better policy loss, but the difference is really minor. Trained beta variables are very close to zero. On average beta seems to be about -0.005 in residual layers.

Low learning rate and starting from already trained weights probably affected the results, but I can't run any longer runs on my computer.

Also earlier I wrote that batch norm mean can be modified to include beta. It's true, but the drawback with that is that net_to_model can't restore the model exactly anymore. At least internally it would be necessary to store the beta variables in weight file too.

**Comment by isty2e** [comment]
@Ttl Then we can safely set center=False, I presume? By the way, what was the reason behind the choice of the training/test set, instead of splitting one dataset into two?

**Comment by Ttl** [comment]
The chunking was done using the old code with games distributed across chunks. Splitted dataset might have different positions from same games in both training and test set which would affect mse loss.

**Comment by lightvector** [comment]
> One small difference I noticed that in minigo L2 regularization is not applied to the bias terms. https://github.com/tensorflow/minigo/blob/188fb197fdafbe9664a32142373b1cbd1459bc67/dual_net.py#L268

> Reading around it seems that generally it's recommended not to apply regularization to biases since large biases don't cause overfitting in the same way as having many non-zero weights. No idea how significant it is though.

@Ttl Don't forget that in the presence of batch normalization (if your activation functions are homogeneous, which relus are), L2 regularization applied to any of the layers before a subsequent batch norm layer has actually no regularization effect, but rather has the effect of bounding the decay of the effective learning rate (assuming I'm remembering/thinking about this correctly). So one might expect L2 regularization to have no bearing on the question of overfitting or not for any layers, either positive or negative, except for the policy head.

**Comment by jkiliani** [comment]
@gcp: I was just made aware by Tobias Berben from German Go platform http://www.go-baduk-weiqi.de/ that Google cloud TPUs have just entered Beta: https://cloudplatform.googleblog.com/2018/02/Cloud-TPU-machine-learning-accelerators-now-available-in-beta.html. Would this be of relevance to the project, at the rates they are charging? (i.e. is it cheaper than Amazon Cloud?)

**Comment by amj** [comment]
@jkiliani i have not yet had luck running MCTS at speed on them, but we hope to get that running this quarter.

For training, it's possible they will help speed things up substantially, but (IIUC) training is not currently the bottleneck (for either LZ or Minigo)

**Comment by gcp** [comment]
Not interesting at all for me.

**Comment by herazul** [comment]
Google's TPU are not interesting for selfplay games ? Would'nt it generate a lot of self play games if LZ code were adapted to run on TPU ?

**Comment by sethtroisi** [comment]
the two teams have regular communication (mostly in discord) so ideas are being shared. Closing this old list as most are now obsolete
